### ISSUE TITLE
Investigation on SVG accessibility

### ISSUE THREAD

**Issue by jessiehuff** [patternfly/patternfly-react#4135]
**Describe the issue. What is the expected and unexpected behavior?**
On the insights dashboard where there are a lot of charts, the charts have a `<title>` which comes up when hovering over the SVG for an extended period of time. The insights team found this to be a usability issue in that the title tooltip that appears can block data in some cases.
![image (1) (1) (2)](https://user-images.githubusercontent.com/28281340/80411554-5f187180-889a-11ea-977f-95df097f493e.png)

In our charts, the aria-title prop is a pass through to the Victory library. The actual Victory property is just title, but we call it aria-title to be clear. The charts are SVGs, so Victory outputs this as a title tag within the svg itself. When you have a title tag, it is native browser behavior to have a tooltip that appears when the mouse is held over a spot on the SVG for an extended period of time. This tooltip appears wherever the mouse is held on the SVG. Taking off the title tag could have potentially negative accessibility side effects.

When researching, I found this article that explains all the different combinations that can be done with SVGs and how they work across different screen readers which provides excellent guidance with the accessibility: https://www.deque.com/blog/creating-accessible-svgs/
Based on the research, it appears that the most recommended combination is `<svg> + role="img" + <title> + <desc> + aria-labelledby="[ID]"` as it was the most reliable choice for the different browser and screen readers that were tested. As the above article explains, taking off the title is not preferred for most use cases as it can cause the SVG to be significantly less accessible. You can remove the title element and add an aria-label like in Pattern 3 in this research, however it will cause the SVG to be read as an image so the user will be unable to interact as easily with the chart by screen reader.

We should investigate if there are improvements that could be made to our charts/SVG accessibility and see if there are better solutions to the title tooltip concerns.

**Please provide the steps to reproduce. Feel free to link CodeSandbox or another tool.**

If you go to our pie chart example and hover over any section of the chart, you can see the title tooltip appear. This appears to be the concern that Insights is having: https://www.patternfly.org/v4/documentation/react/charts/chartpie/basic-with-right-aligned-legend

**Is this a bug or enhancement? If this issue is a bug, is this issue blocking you or is there a work-around?**
Enhancement to our charts accessibility.
Work around - if you see this as a significant usability issue for visual users, you can take off the title as long as it's replaced with something else like a proper aria-desc to ensure the user is getting the information. You could use an aria-label, but you have to be aware of the fact that it will make the svg be read as an image and be less accessible. You would need to relay the information inside the chart in a different way to screen reader users. The above article explains all of the different combinations that you could try and the different results you will have given the pattern you choose.

**Comment by dlabrecq** [comment]
A while back, we performed an a11y investigation regarding charts. Victory charts is already doing what the article above suggests. That is, title and description are supported, as well as providing aria roles.

See https://github.com/patternfly/patternfly-react/issues/1498 and https://github.com/patternfly/patternfly-react/issues/2079

>Victory charts support a11y properties for description and title.
>
>https://formidable.com/open-source/victory/docs/common-container-props/#desc
>https://formidable.com/open-source/victory/docs/common-container-props/#title
>
>Victory also provides aira roles as well:
>https://github.com/FormidableLabs/victory-chart/pull/233

**Comment by dlabrecq** [comment]
We cannot change how the browser chooses to display the title. However, the app developer can override the tooltip component's position.

For example, the tooltip can be shifted horizontally and/or vertically shift from the x / y coordinates using the `dx` and `dy` props.

See https://formidable.com/open-source/victory/docs/victory-tooltip#dx and
https://formidable.com/open-source/victory/docs/victory-tooltip#dy

The `x` any `y` props define the x / y coordinates to use as a basis for positioning the tooltip element.

See https://formidable.com/open-source/victory/docs/victory-tooltip#x and
https://formidable.com/open-source/victory/docs/victory-tooltip#y

There are also properties such as `center`, `centerOffset`, etc. that could be useful to developers.

See https://formidable.com/open-source/victory/docs/victory-tooltip#centeroffset and
https://formidable.com/open-source/victory/docs/victory-tooltip#center

**Comment by dlabrecq** [comment]
A while back, we performed an a11y investigation regarding charts. See https://github.com/patternfly/patternfly-react/issues/1498 and https://github.com/patternfly/patternfly-react/issues/2079

Victory charts already render what the blog suggests above and more. That is, `title` and `description` are supported, as well as providing aria roles such as `img`, `presentation`, etc. for various tags (e.g., `svg`, `g`, etc.).

>You can remove the title element and add an aria-label like in Pattern 3

Note that Pattern 3 refers to using an SVG with an `img` tag. What Victory outputs is more similar to Pattern 10.

That said, omitting the `title` may be an option when using the `desc` property. The `desc` prop appears to work with Mac VoiceOver and IE11 on Windows with JAWS.

In general, SVGs are not accessible because it's still just an image. Having both `title` and `desc` are probably not as useful as providing a table to represent the chart's data.

**Comment by dlabrecq** [comment]
We cannot change how the browser chooses to display the `title`. However, the app developer can override the tooltip component's position.

For example, the tooltip can be shifted horizontally and vertically from the x / y coordinates using the `dx` and `dy` props.

See the https://formidable.com/open-source/victory/docs/victory-tooltip#dx and
https://formidable.com/open-source/victory/docs/victory-tooltip#dy examples.

The tooltip's `x` any `y` props define the x / y coordinates to use as a basis for positioning the tooltip element.

See the https://formidable.com/open-source/victory/docs/victory-tooltip#x and
https://formidable.com/open-source/victory/docs/victory-tooltip#y examples.

There are also properties such as `center`, `centerOffset`, etc. that could be useful to developers.

See the https://formidable.com/open-source/victory/docs/victory-tooltip#centeroffset and
https://formidable.com/open-source/victory/docs/victory-tooltip#center examples.

**Comment by dlabrecq** [comment]
The pie chart in the above screenshot appears to be much smaller than our examples. However, I was able to create a very small pie chart (100px in height) with a single dataset. 

Note: This was not reproducible with multiple datasets. I'm also using `constrainToVisibleArea={true}`

In the example below, the cursor is moved to show when the `title` appears. Only when I move the cursor to the very bottom, can I get the `title` to appear above the tooltip. However, the tooltip is easily re-positioned simply by moving the cursor.

![Screen-Recording-2020-04-27-at-10 49 57-PM](https://user-images.githubusercontent.com/17481322/80442394-cbb36080-88da-11ea-975c-9a2b6725e3ff.gif)

**Comment by christiemolloy** [comment]
Adding to this conversation about SVG accessibility, I have been investigating how to get the screenreader to read the information in the chart. Currently PF charts wrap around the Victory chart which is where the svg properties are set. We cannot override any of these. Which means that we cannot override `<svg role="img">`. With `role=”img”` set on the `<svg>`, the screen reader will not read any nested elements inside of this. E.g. if you want to set `role="table"` or `role="list"` inside of the `svg` the screenreader will not read those.

After talking with @jessiehuff and @dlabrecq , these are two solutions which we thought could be viable options so I tested them out.

**Solution 1:**  Overrides the roles.

[charts-example-1-small.mov.zip](https://github.com/patternfly/patternfly-react/files/4595177/charts-example-1-small.mov.zip)

Roles that it uses:
`<svg role=”table”>`  would override the role=”img” on the svg
`<text role=”row”>`
`<tspan role=”cell”>`

**Pros:** 
- It will read out all contents in the chart legend as you tab through

**Cons:** 
- Adding roles only applies to text and tspan elements which come with the chart legend, but not the chart. So if you are using a chart without a legend you cannot apply these roles.
- We would need to see if Victory would be open to passing through different props for the role set on svg, rather than only having role=”img”

**Solution 2:** HTML table markup that was written additionally to the chart
```
<table tabIndex=”0” className=”visually-hidden” aria-label=”Patch chart data”
  <tr>
    <td>246</td>
    <td>security advisories</td>
  </tr>
  <tr>
    <td>712</td>
    <td>bug fixes</td>
  </tr>
  <tr>
    <td>88</td>
    <td>enhancements</td>
  </tr>
</table>
```
```
.visually-hidden {
   position: absolute !important;
   top: -40px;
   font-size: 10px;
   height: 1px;
   width: 1px;
   overflow: hidden;
   clip: rect(1px 1px 1px 1px); /* IE6, IE7 */
   clip: rect(1px, 1px, 1px, 1px);
   white-space: nowrap; /* added line */
}
```

**Pros:** 
- reads out all elements in the table
- We don’t need to override anything in Victory
- This will work for charts with and without a legend

**Cons:**
- Extra markup for each chart (but this could come with the chart)
- Tabs through a hidden table, vs the roles which tab through the actual chart legend (but i don’t think this is a big con)

**Comment by dlabrecq** [comment]
Adding roles helps read the chart legend labels, but not the chart data. Although, filing an issue with Victory is worthwhile, providing a PR would be even better.

The hidden table approach makes the raw data available to a screen reader. It would be ideal if users could walk through the data presented by the chart. This would be truly accessible -- it's something we've been talking about for a while.

The hidden table would be particularly useful for QA testing.

**Comment by dlabrecq** [comment]
Adding roles helps read the chart legend labels, but not the chart data. (Not sure how helpful it is to read a legend which is disconnected from chart data — unless the label contains data itself?) Although, filing an issue with Victory is still worthwhile, and providing a PR would be even better.

It would be ideal if screen reader users could walk through the data presented by the chart. The hidden table approach makes the raw data available in a truly accessible way -- it's something we've been talking about for a while.

For a previous app I worked on, I implemented accessibility and high contrast modes. When enabled, tables would be displayed instead of charts. IMO, that is really the best way to make chart data accessible.

In addition, the hidden table would be particularly useful for QA testing. That is, tables could be used to confirm a chart's expected data.

**Comment by christiemolloy** [comment]
Agreed @dlabrecq I'll still file an issue/create a PR with Victory to see if they would be open to letting us pass through a prop to set the `role` so that will give us flexibility in the future. It also means that if the legend is disconnected from the chart like in the case of the insights dashboard, it wouldn't read out that the image doesn't have any labels/captions.

And yes I think the table works great, do you think that this should be contributed back to pf4 so that all the charts have a default table for accessibility? If @jessiehuff also agrees with the table approach then I can create a follow-up issue and add that.

I think seeing the table for high contrast modes is a good idea. Maybe that would be a case by case basis though. E.g. the user could pass through a prop to disable the hidden styles of the table at a high contrast mode?

+1 to QA testing.

**Comment by dlabrecq** [comment]
It would be interesting to see how to represent various charts as a table. It may be a challenge to represent thresholds with some of the more complex charts (e.g., donut utilization, bullet chart, etc.). I'd love to see us put some effort behind this, tho.

**Comment by dgutride** [comment]
@christiemolloy  - can you put a link to the victory issue in this issue here once it's created, pls?

**Comment by christiemolloy** [comment]
@dgutride PR is here https://github.com/FormidableLabs/victory/pull/1643
